[Unified PDF] Jetsams hit when quickly scrolling via Find-in-page results

aprotyas opened the pull request:
#### 5c8e008844109a2b2cf1bdc1819f50f4b9a41a1a
<pre>
[Unified PDF] Jetsams hit when quickly scrolling via Find-in-page results
<a href="https://bugs.webkit.org/show_bug.cgi?id=294617">https://bugs.webkit.org/show_bug.cgi?id=294617</a>
<a href="https://rdar.apple.com/149223603">rdar://149223603</a>

Reviewed by Aditya Keerthi.

We notice the web content process often jetsams when we there is a large
number find-in-page results and we jump through them. This is explained
by unnecessary memory incurred in:

1. UnifiedPDFPlugin::visibleRectsForFindMatchRects(): Where we provide
   rects for _every_ find match, even if they are not within the
   unobscured content rect. Thus, WebFoundTextRangeController::drawRect
   paints many more paths than it needs to.
2. UnifiedPDFPlugin::selectionFromWebFoundTextRangePDFData(): Where we
   call into -[PDFDocument selectionFromPage:...] for every find-in-page
   result. This produces O(n) memory allocations under
   CGPDFSelectionCreateForRange.

Fixes are explained in line.

* Source/WebKit/Shared/WebFoundTextRange.cpp:
(WebKit::WebFoundTextRange::PDFData::hash const):
(WebKit::WebFoundTextRange::hash const):
(WebKit::operator&lt;&lt;):
    While the patch does not add new logging of PDFData, this is a useful
    utility to have.
* Source/WebKit/Shared/WebFoundTextRange.h:
(WTF::WebFoundTextRangePDFDataHash::hash):
(WTF::WebFoundTextRangePDFDataHash::equal):
(WTF::HashTraits&lt;WebKit::WebFoundTextRange::PDFData&gt;::emptyValue):
(WTF::HashTraits&lt;WebKit::WebFoundTextRange::PDFData&gt;::isEmptyValue):
(WTF::HashTraits&lt;WebKit::WebFoundTextRange::PDFData&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebKit::WebFoundTextRange::PDFData&gt;::isDeletedValue):
    Provide hash traits for PDFData, allowing it to be used as a map key.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::releaseMemory):
(WebKit::UnifiedPDFPlugin::findString):
     Remember to release the cache memory when appropriate, i.e. when a
     find session terminates or when we are generally releasing memory
     in the plugin.
(WebKit::UnifiedPDFPlugin::visibleRectsForFindMatchRects const):
    Only return rects for the find matches that currently intersect the
    unobscured content rect. This means we don&apos;t have an O(n) number of
    find match rects to paint.
(WebKit::UnifiedPDFPlugin::selectionFromWebFoundTextRangePDFData):

Canonical link: <a href="https://commits.webkit.org/296356@main">https://commits.webkit.org/296356@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/637fd71279086a91254030d669738c8223f95dbb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27926 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113474 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28620 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36477 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82202 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111213 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97526 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62635 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22099 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58207 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92048 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15718 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91226 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35696 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91023 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23200 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35915 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13685 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31085 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35224 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40766 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34946 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36622 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1438,6 +1438,8 @@ static String mutationObserverNotificationString()
     if (m_presentationController)
         m_presentationController->releaseMemory();
+
+    webFoundTextRangePDFDataSelectionMap().clear();
```
We should probably clear when a search session ends too?

aprotyas replied:
Good point

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/d8e481947debb24022d7e6d5cf4477cbc3c0f5db!

pxlcoder commented on Source/WebKit/Shared/WebFoundTextRange.h:
```diff
@@ -57,10 +62,43 @@ struct WebFoundTextRange {
+template<> struct HashTraits<WebKit::WebFoundTextRange::PDFData> : GenericHashTraits<WebKit::WebFoundTextRange::PDFData> {
+    static constexpr bool emptyValueIsZero = false;
+    static constexpr bool needsDeletedValue = true;
+private:
```
Nit: I think we normally do a newline before the access declaration.

pxlcoder commented on Source/WebKit/Shared/WebFoundTextRange.h:
```diff
@@ -57,10 +62,43 @@ struct WebFoundTextRange {
+private:
+    static constexpr auto sentinel { std::numeric_limits<uint64_t>::max() };
+    static constexpr auto deletedSentinel { sentinel - 1 };
+public:
```
Ditto.

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3386,17 +3389,33 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
 
+HashMap<WebFoundTextRange::PDFData, RetainPtr<PDFSelection>>& UnifiedPDFPlugin::webFoundTextRangePDFDataSelectionMap()
+{
+    static NeverDestroyed<HashMap<WebFoundTextRange::PDFData, RetainPtr<PDFSelection>>> map;
```
Wouldn't it make more sense for this to be per-instance rather than static?

aprotyas replied:
I think we'd get similar performance characteristics either way (as long as we manage the cache properly), but you're correct that conceptually this map belongs to a single Unified PDF instance.

aprotyas replied:
Let me change this map to a data member.

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3252,6 +3254,7 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
         m_lastFindString = target;
         setCurrentSelection(nullptr);
         m_findMatchRects.clear();
+        webFoundTextRangePDFDataSelectionMap().clear();
```
This will clear the map when a new search is performed, but if you have one in progress and then close the find bar, we'd still be keeping around the cache contents right? Is that okay?

aprotyas replied:
We wouldn't be, because closing the find bar calls into `hideFindUI()` too, which calls `UnifiedPDFPlugin::findString("", ...)`

aprotyas replied:
So this should be fine

pxlcoder replied:
Nice!

pxlcoder reviewed (approved):
r+ after making the map a member variable.

aprotyas committed:
[Unified PDF] Jetsams hit when quickly scrolling via Find-in-page results
https://bugs.webkit.org/show_bug.cgi?id=294617
rdar://149223603

Reviewed by Aditya Keerthi.

We notice the web content process often jetsams when we there is a large
number find-in-page results and we jump through them. This is explained
by unnecessary memory incurred in:

1. UnifiedPDFPlugin::visibleRectsForFindMatchRects(): Where we provide
   rects for _every_ find match, even if they are not within the
   unobscured content rect. Thus, WebFoundTextRangeController::drawRect
   paints many more paths than it needs to.
2. UnifiedPDFPlugin::selectionFromWebFoundTextRangePDFData(): Where we
   call into -[PDFDocument selectionFromPage:...] for every find-in-page
   result. This produces O(n) memory allocations under
   CGPDFSelectionCreateForRange.

Fixes are explained in line.

* Source/WebKit/Shared/WebFoundTextRange.cpp:
(WebKit::WebFoundTextRange::PDFData::hash const):
(WebKit::WebFoundTextRange::hash const):
(WebKit::operator<<):
    While the patch does not add new logging of PDFData, this is a useful
    utility to have.
* Source/WebKit/Shared/WebFoundTextRange.h:
(WTF::WebFoundTextRangePDFDataHash::hash):
(WTF::WebFoundTextRangePDFDataHash::equal):
(WTF::HashTraits<WebKit::WebFoundTextRange::PDFData>::emptyValue):
(WTF::HashTraits<WebKit::WebFoundTextRange::PDFData>::isEmptyValue):
(WTF::HashTraits<WebKit::WebFoundTextRange::PDFData>::constructDeletedValue):
(WTF::HashTraits<WebKit::WebFoundTextRange::PDFData>::isDeletedValue):
    Provide hash traits for PDFData, allowing it to be used as a map key.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::releaseMemory):
(WebKit::UnifiedPDFPlugin::findString):
     Remember to release the cache memory when appropriate, i.e. when a
     find session terminates or when we are generally releasing memory
     in the plugin.
(WebKit::UnifiedPDFPlugin::visibleRectsForFindMatchRects const):
    Only return rects for the find matches that currently intersect the
    unobscured content rect. This means we don't have an O(n) number of
    find match rects to paint.
(WebKit::UnifiedPDFPlugin::selectionFromWebFoundTextRangePDFData):

Canonical link: https://commits.webkit.org/296356@main
